feat: Import and export from ARK CLI

skazinka opened the pull request:
## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person> for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

skazinka committed:
feat: export command

to export ARK resources to yaml. uses kubectl get to get the resources and writes them to a file

skazinka committed:
feat: import command

currently just a thin wrapper around kubectl create, but it seems polite to have an import command if there's an export command

skazinka committed:
fix: check for correct output type in test

and check that we export in dependency order regardless of what order the types are specified

skazinka committed:
add test for import, export failure

tweaks export tool behavior to fail the first time it encounters an ungettable resource; we don't want an accidental partial export to be mistaken for a full one.

skazinka committed:
Merge branch 'main' into feat/cli-export

dwmkerr commented on tools/ark-cli/src/commands/export/index.ts:
```diff
@@ -0,0 +1,136 @@
+  labels?: string;
+}
+
+async function getResources(
```
can we look at the other kubectl commands we have put them into a separate folder, we can do the same here

skazinka replied:
Ah, cool, I didn't spot that. Now uses the shared listResources

dwmkerr commented on tools/ark-cli/src/index.tsx:
```diff
@@ -59,7 +61,9 @@ async function main() {
   program.addCommand(createEvaluationCommand(config));
+  program.addCommand(createExportCommand(config));
   program.addCommand(createGenerateCommand(config));
+  program.addCommand(createImportCommand(config));
```
ask claude/cursor to update shell expansion command then it works with `ark <tab>`

skazinka replied:
Done 👍

dwmkerr commented on tools/ark-cli/src/commands/export/index.ts:
```diff
@@ -0,0 +1,136 @@
+import output from '../../lib/output.js';
+
+const ARK_RESOURCE_TYPES = [
+  'secrets',
```
i think we should put these in the `config` object w/ defaults then comment in `sample config yaml` how to customise - cause then people can add/remove their own preferences and it keeps it nicely separate

skazinka replied:
Good idea. I do still have an array in the code because it's necessary to be able to enforce order - if, for example, the teams are exported before the agents they contain, then importing won't work. But now the default behavior can be overridden (though not the ordering).

dwmkerr replied:
Then i think that this is in the `config` at https://github.com/mckinsey/agents-at-scale-ark/pull/536/files#diff-d04dac0d506afa03936b01bc13586f8610ad4044fc9186dc6e000070ba56e3a1L30 i.e. the config object handles the defaults

dwmkerr reviewed (changes requested):
minor changes!

skazinka committed:
Fix export test mocks to correct format

Now tests that we're actually trying to save, which we weren't when there was no data from the mock

skazinka committed:
Update ark completions

skazinka committed:
Using listResources from kubectl.ts

Reusing existing kubectl function, by request. Requires turning the response into yaml rather than having kubectl do it, and modifying shared function to also take namespace and labels into account.

skazinka committed:
Allow overriding default export types in config

For when there is no -t specified.
It's still necessary to have the order in which to export specified in the file, because if the resources are exported out of order they may be able to be imported. I.e. a team cannot be imported before the agents that it depends on exist.

dwmkerr commented on tools/ark-cli/.arkrc.yaml.sample:
```diff
@@ -38,3 +38,14 @@ services:
       - containerRegistry.password=YOUR_PASSWORD
+
+# Which resources ark export should export if not otherwise specified
+defaultExportTypes:
```
@Lily Kazinka for the config file, the defaultExportTypes would actually be `exportTypes` wouldn't it? cause basically they're what you use to override the defaults, am i correct they are an array and that's why you've got it like that to get the order right, a bit like queryTimeout is not defaultQueryTimeout, it's actually the value to _override_ the default

I always forget with yaml - the below is an array right, so you get the right order, if so would be useful in the comment

skazinka replied:
Hm, I guess that's true. I was thinking "default" in that it's what you get if you don't specify which types you want (i.e. if you run `ark export` rather than `ark export -t foo,bar`), so it's still the default from the user's perspective. I can change the name if you think it's confusing though.

I wanted to make sure that users can't specify an order in the config that could produce unusable exports, so while this is an ordered data structure, it won't affect the order of the actual export which is checked against the hardcoded order. I can remove that check if you prefer.

dwmkerr replied:
Thanks for explaining! makes sense :)

dwmkerr reviewed (dismissed):
sorry minor changes! I think we also have an `ark config` command we can enrich with this - it just show the current config so would be a quick way to test it too

skazinka committed:
Merge branch 'main' into feat/cli-export

skazinka committed:
Merge branch 'main' into feat/cli-export

skazinka committed:
Merge branch 'main' into feat/cli-export

skazinka committed:
Merge branch 'main' into feat/cli-export

skazinka committed:
Merge branch 'main' into feat/cli-export
